pkg/process/process.go

```diff
@@ -6,45 +6,92 @@ package process
 import (
 	"context"
 	"flag"
+	"log"
 
-	"github.com/google/uuid"
+	"github.com/spacemonkeygo/flagfile"
+	"github.com/zeebo/errs"
 	"go.uber.org/zap"
 	monkit "gopkg.in/spacemonkeygo/monkit.v2"
 
+	"storj.io/storj/pkg/telemetry"
 	"storj.io/storj/pkg/utils"
 )
 
+var (
+	logDisposition = flag.String("log.disp", "prod",
+		"switch to 'dev' to get more output")
+
+	// Error is a process error class
+	Error = errs.Class("ProcessError")
+)
+
 // ID is the type used to specify a ID key in the process context
 type ID string
 
 // Service defines the interface contract for all Storj services
 type Service interface {
+	// Process should run the program
 	Process(context.Context) error
+
 	SetLogger(*zap.Logger) error
 	SetMetricHandler(*monkit.Registry) error
+
+	// InstanceID should return a server or process instance identifier that is
+	// stable across restarts, or the empty string to use the first non-nil
+	// MAC address
+	InstanceID() string
 }
 
-var (
+const (
 	id ID = "SrvID"
 )
 
 // Main initializes a new Service
-func Main(s Service) error {
-	flag.Parse()
+func Main(s Service) (err error) {
+	flagfile.Load()
+
 	ctx := context.Background()
-	uid := uuid.New().String()
 
-	logger, err := utils.NewLogger("", zap.Fields(zap.String("SrvID", uid)))
+	instanceID := s.InstanceID()
+	if instanceID == "" {
+		instanceID = telemetry.DefaultInstanceID()
+	}
+
+	ctx, cf := context.WithCancel(context.WithValue(ctx, id, instanceID))
+	defer cf()
+
+	registry := monkit.Default
+	scope := registry.ScopeNamed("process")
+	defer scope.TaskNamed("main")(&ctx)(&err)
+
+	logger, err := utils.NewLogger(*logDisposition,
+		zap.Fields(zap.String(string(id), instanceID)))
 	if err != nil {
 		return err
 	}
 	defer logger.Sync()
-
-	ctx, cf := context.WithCancel(context.WithValue(ctx, id, uid))
-	defer cf()
+	defer zap.ReplaceGlobals(logger)()
```

Comment:
Why do we need this?

Reply:
this is so we can use zap.L() during debugging without having to pass a logger all the way in. i would prefer people occasionally use a global logger instead of just using fmt.Printf, which injects debug data into stdout.